src/realm/object-store/sync/impl/emscripten/socket_provider.cpp

```diff
@@ -0,0 +1,219 @@
+        : m_handler(std::move(handler))
+        , m_timeout(emscripten_set_timeout(timeout_callback, doubleMiliseconds(delay).count(), this))
+        , m_scheduler(scheduler)
+    {
+    }
+    virtual ~EmscriptenTimer() final
+    {
+        cancel();
+    }
+    virtual void cancel() final
+    {
+        if (m_timeout && m_handler) {
+            emscripten_clear_timeout(*m_timeout);
+            m_scheduler.invoke([handler = std::move(m_handler)]() {
+                handler(Status(ErrorCodes::OperationAborted, "Timer canceled"));
+            });
+        }
+    }
+
+private:
+    SyncSocketProvider::FunctionHandler m_handler;
+    std::optional<int> m_timeout;
+    util::EmscriptenScheduler& m_scheduler;
+
+    static void timeout_callback(void* user_data)
+    {
+        auto timer = reinterpret_cast<EmscriptenTimer*>(user_data);
+        timer->m_timeout = std::nullopt;
+        timer->m_handler(Status::OK());
+    }
+};
+
+struct EmscriptenWebSocket final : public WebSocketInterface {
+public:
+    EmscriptenWebSocket(EMSCRIPTEN_WEBSOCKET_T socket, std::unique_ptr<WebSocketObserver> observer)
+        : m_socket(socket)
+        , m_sentinel(std::make_shared<LivenessSentinel>())
+        , m_observer(std::move(observer))
+    {
+        emscripten_websocket_set_onopen_callback(m_socket, m_observer.get(), open_callback);
+        emscripten_websocket_set_onmessage_callback(m_socket, m_observer.get(), message_callback);
+        emscripten_websocket_set_onerror_callback(m_socket, m_observer.get(), error_callback);
+        emscripten_websocket_set_onclose_callback(m_socket, m_observer.get(), close_callback);
+    }
+
+    virtual ~EmscriptenWebSocket() final
+    {
+        emscripten_websocket_close(m_socket, 0, nullptr);
+        emscripten_websocket_delete(m_socket);
+    }
+
+    virtual void async_write_binary(util::Span<const char> data, SyncSocketProvider::FunctionHandler&& handler) final
+    {
+        emscripten_websocket_send_binary(m_socket, const_cast<char*>(data.data()), data.size());
+
+        auto weak_handler = [handler = std::move(handler), sentinel = std::weak_ptr(m_sentinel)](Status status) {
+            if (sentinel.lock()) {
+                // only call the real handler if the socket object is still alive
+                handler(status);
+            }
+        };
+        emscripten_set_timeout(sending_poll_check, 0, new PollCheckState{m_socket, std::move(weak_handler)});
+    }
+
+private:
+    struct PollCheckState {
+        EMSCRIPTEN_WEBSOCKET_T socket;
+        SyncSocketProvider::FunctionHandler handler;
+        double next_delay = 1;
+    };
+
+    struct LivenessSentinel {
+    };
+
+    static void sending_poll_check(void* user_data)
+    {
+        std::unique_ptr<PollCheckState> state(reinterpret_cast<PollCheckState*>(user_data));
+        unsigned long long buffered_amount;
+        emscripten_websocket_get_buffered_amount(state->socket, &buffered_amount);
```

Comment:
Looking at https://github.com/emscripten-core/emscripten/blob/bc3100472cc01d66f052aafa9a14fb61717a12f2/src/library_websocket.js#L15 and https://github.com/emscripten-core/emscripten/blob/bc3100472cc01d66f052aafa9a14fb61717a12f2/src/library_websocket.js#L30 it looks like the only error condition is the socket file descriptor being completely made up, so I didn't think there's anything we can handle here. I could add a `REALM_ASSERT` on the result if you'd like, but I'm not sure we'd ever hit it.

Reply:
I get nervous whenever I see an error being potentially ignored. C-style return codes are terrible for this because that is the invisible (in code) default if you do nothing else. So it is impossible to distinguish between intentionally and accidentally ignoring errors.

Ideally EMSCRIPTEN_RESULT would have been tagged with `[[nodiscard]]` or its compiler-specific version for C so that we would get warnings if we ignored it. I'd suggest behaving as-if it were marked and check every result, even if we don't think they could ever happen. You could even consider temporarily modifying your local copy of the emscripten header to add that annotation so that we can find all missing checks.